### PR TITLE
Use revamped topology APIs in topotree

### PR DESCRIPTION
**Pull request by gcongiu**
## Pull Request Description

This PR depends from #4127 and should not be merged before it.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by jain-surabhi-23** [comment]
@gcongiu By the first look, the code looks fine to me. Just to double check, can you please do a similar run as last time to make sure the topology tree generated it correct?

**Comment by gcongiu** [comment]
> @gcongiu By the first look, the code looks fine to me. Just to double check, can you please do a similar run as last time to make sure the topology tree generated it correct?

Sure

**Comment by gcongiu** [comment]
@jain-surabhi-23 I have fixed a compatibility problem that arose because the revamped topology layer uses global id while your implementation uses `hwloc` logical indices to represent objects in the tree. With this fix the new implementation gives same results as the old one.

File: src/include/mpir_hwtopo.h [hunk 1]
```diff
@@ -57,6 +57,14 @@ bool MPIR_hwtopo_is_initialized(void);
...
 MPIR_hwtopo_gid_t MPIR_hwtopo_get_leaf(void);
 
+/*
+ * Return the local id for the object with global id: gid. The local
+ * id corresponds to the hwloc's logical index of objects at a certain
+ * level of the topology tree. This means that if there are N objects
+ * at one level, the logical index is an integer in the range [0:N)
+ */
+int MPIR_hwtopo_get_lid(MPIR_hwtopo_gid_t gid);
+
```
Comment (gcongiu) [inline src/include/mpir_hwtopo.h]: @hzhou I am not sure you will like this but I wonder if it is ok for now as reimplementing the topotree code to make it hwloc logical index independent is much more work

**Comment by jain-surabhi-23** [comment]
I am out of office for 2 days. I will get back to you on Friday or early next week.

**Comment by hzhou** [comment]
@gcongiu This one looks good to me. You may add my approval as well.

**Pull request merged**